### PR TITLE
Library rewrite to support additional features

### LINKED ISSUE #9: Update readme with setup and usage examples

### PR DESCRIPTION
**Pull request by bogdanromanx**
Library rewrite to support additional features

Significant changes:
-   new implementation for non-clustered deployments
-   asynchronous evaluation of commands
-   support for snapshots
-   support for describing arbitrary passivation strategies
-   support for describing arbitrary retry strategies
-   generic implementation in `F[_]` using cats-effect typeclasses

Fixes #9, #19, #20.

### CODE REVIEW

**Review by hygt — APPROVED** [review]

File: modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/AggregateActor.scala [hunk 1]
```diff
@@ -1,121 +1,493 @@
...
+                persistenceId,
+                Event.runtimeClass.getSimpleName)
+    // $COVERAGE-ON$
   }
 
   override def receiveCommand: Receive = {
-    case Append(id, value) =>
-      Event.cast(value) match {
-        case Some(event) =>
+    case Append(mid, value) if mid == id =>
```
Comment (umbreak) [inline modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/AggregateActor.scala:104]:
```scala
case Append(`id`, value) =>
```

File: modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/AggregateActor.scala [hunk 2]
```diff
@@ -1,121 +1,493 @@
...
-              log.debug("Rejected command '{}' on actor '{}' because '{}'", cmd, persistenceId, rejection)
-              sender() ! Evaluated(id, result)
-            case Right(event) =>
-              persist(event) { _ =>
-                log.debug("Applied event '{}' to actor '{}'", event, persistenceId)
-                state = next(state, event)
-                sender() ! Evaluated(id, Right(state))
-              }
-          }
+    case GetLastSeqNr(mid) if mid == id =>
```
Comment (umbreak) [inline modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/AggregateActor.scala:122]: same

File: modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/AggregateActor.scala [hunk 3]
```diff
@@ -1,121 +1,493 @@
...
+      sender() ! LastSeqNr(id, lastSequenceNr)
+      log.debug("Replied with LastSeqNr '{}' from actor '{}'", lastSequenceNr, persistenceId)
+      passivateAfterEvaluation()
+    case GetCurrentState(mid) if mid == id =>
```
Comment (umbreak) [inline modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/AggregateActor.scala:126]: same

File: modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/AggregateActor.scala [hunk 4]
```diff
@@ -1,121 +1,493 @@
...
+      sender() ! CurrentState(id, state)
+      log.debug("Replied with CurrentState '{}' from actor '{}'", state, persistenceId)
+      passivateAfterEvaluation()
+    case Evaluate(mid, value) if mid == id =>
```
Comment (umbreak) [inline modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/AggregateActor.scala:130]: same

File: modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/AggregateActor.scala [hunk 5]
```diff
@@ -1,121 +1,493 @@
...
+          sender() ! TypeError(id, Command.runtimeClass.getSimpleName, value)
+          passivateAfterEvaluation()
         // $COVERAGE-ON$
       }
-    case CheckEval(id, value) =>
-      Command.cast(value) match {
-        case Some(cmd) =>
-          sender() ! Validated(id, eval(state, cmd).left.toOption)
+
+    case Test(mid, value) if mid == id =>
```
Comment (umbreak) [inline modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/AggregateActor.scala:146]: same

File: modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/AggregateActor.scala [hunk 6]
```diff
@@ -1,121 +1,493 @@
...
+        case _ =>
+          log.error("Received a command '{}' incompatible with the expected type '{}'",
+                    value,
+                    Command.runtimeClass.getSimpleName)
+          sender() ! TypeError(id, Command.runtimeClass.getSimpleName, value)
+          passivateAfterEvaluation()
         // $COVERAGE-ON$
       }
-    case GetLastSeqNr(id) =>
+    case Snapshot(mid) if mid == id =>
```
Comment (umbreak) [inline modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/AggregateActor.scala:161]: same

**Review by umbreak — CHANGES_REQUESTED** [review]

File: modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/RetryStrategy.scala [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    * @param initialDelay the initial delay to apply before retrying the computation
+    * @param maxRetries   the maximum number of tries
+    * @param factor       the delay increase factor in between retries
+    * @param T            a timer for __F__
+    * @param F            an applicative error for __F__
+    * @tparam F           the effect type
+    * @tparam E           the error type
+    * @return a retry strategy with exponential backoff
+    */
+  def exponentialBackoff[F[_], E](
```
Comment (umbreak) [inline modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/RetryStrategy.scala:74]: [Exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) is not deterministic, being the time between retries selected among a random range of values as: `2^c − 1`.
You might just want to call the method `exponential` if you just want to double the waiting time per retry.

**Review by umbreak — CHANGES_REQUESTED** [review]

**Review by umbreak — APPROVED** [review]

File: modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/RetryStrategy.scala (hunk 7 above)
Comment (bogdanromanx) [reply modules/akka/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/akka/RetryStrategy.scala:74]: This seems to be the most common way to refer to what this fn does...

**Pull request merged**